app/models/category_featured_topic.rb

```diff
@@ -60,24 +60,19 @@ def self.feature_topics_for(c, existing = nil)
 
     # Add topics, even if they're in secured categories or invisible
     query = TopicQuery.new(Discourse.system_user, query_opts)
-    results = query.list_category_topic_ids(c).uniq
+    results = query.list_category_topic_ids(c)
 
     # Add some topics that are visible to everyone:
     anon_query = TopicQuery.new(nil, query_opts.merge(except_topic_ids: [c.topic_id] + results))
-    results += anon_query.list_category_topic_ids(c).uniq
+    results += anon_query.list_category_topic_ids(c)
 
+    results.uniq!
     return if results == existing
 
     CategoryFeaturedTopic.transaction do
       CategoryFeaturedTopic.where(category_id: c.id).delete_all
-      if results
-        results.each_with_index do |topic_id, idx|
-          begin
-            c.category_featured_topics.create(topic_id: topic_id, rank: idx)
-          rescue PG::UniqueViolation
-            # If another process features this topic, just ignore it
-          end
-        end
+      results.each_with_index do |topic_id, idx|
+        c.category_featured_topics.create(topic_id: topic_id, rank: idx)
```

Comment:
So if another process features one of the topics, what happens? Before, it would just catch the unique violation exception and keep featuring the other topics. With your version, as soon as there's any sort of errors, the whole transaction is reverted and nothing is featured?

Reply:
All featured topics are deleted first and then recreated here. Because all of this runs in a transaction, there should be no conflicts at all. The previous rescue did not work either because it should have been catching ActiveRecord::RecordNotUnique.

Reply:
Maybe the exception was the right one but then we added a unique contraint and the exception changed?